

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tutorial 6: Two-Step Refinement Process &mdash; chromo  documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  
    <link rel="stylesheet" href="_static/plot_directive.css" type="text/css" />
  
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="chromo  documentation" href="index.html"/>
        <link rel="next" title="Tutorial 7: Kinked Chromatin" href="tutorial_7_kinked_chromatin.html"/>
        <link rel="prev" title="Tutorial 5: Two-Mark Chromatin with Competitive Binding" href="tutorial_5_two_mark_chromatin_competitive.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> chromo
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="abstract.html">Abstract</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_structure.html">Code Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="mc_sim.html">Monte Carlo Simulations</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Theory</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="poly_models.html">Polymer Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="field.html">Field Theoretic Treatment of Interactions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorial_1_unconfined_semiflexible_homopolymer.html">1. Unconfined, Semiflexible Homopolymer</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_2_confined_semiflexible_homopolymer.html">2. Confined, Semiflexible Homopolymer</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_3_one_mark_chromatin.html">3. Single-Mark Chromatin Simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_4_two_mark_chromatin.html">4. Two-Mark Chromatin Simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_5_two_mark_chromatin_competitive.html">5. Two-Mark Chromatin Simulation with Competitive Binding</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">6. Simulation with Coarse and Refined Steps</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Import-Modules">Import Modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Specify-Binders">Specify Binders</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Specify-the-Confinement">Specify the Confinement</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Instantiate-Epigenetic-Mark-Pattern">Instantiate Epigenetic Mark Pattern</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Specify-Initial-Reader-Protein-Binding-States">Specify Initial Reader Protein Binding States</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Instantiate-the-Polymer">Instantiate the Polymer</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Specify-the-Uniform-Density-Field">Specify the Uniform Density Field</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Specify-the-Coarse-Grained-Model">Specify the Coarse-Grained Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Specify-the-Simulation-Parameters">Specify the Simulation Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Run-the-Coarse-Grained-Simulation">Run the Coarse-Grained Simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Plot-the-Coarse-Grained-Chromatin-Fiber">Plot the Coarse-Grained Chromatin Fiber</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Refine-the-Coarse-Grained-Model">Refine the Coarse-Grained Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Important-Consideration">Important Consideration</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id1">Specify the Simulation Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Run-the-Refined-Simulation">Run the Refined Simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Plot-the-Refined-Chromatin-Fiber">Plot the Refined Chromatin Fiber</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_7_kinked_chromatin.html">7. Kinked Chromatin Simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_8_kinked_chromatin_with_sterics.html">8. Kinked Chromatin Simulation with Sterics</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_snippets.html">Code Snippets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Additional Resources</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules.html">Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">chromo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Tutorial 6: Two-Step Refinement Process</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/tutorial_6_two_step_refinement.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Tutorial-6:-Two-Step-Refinement-Process">
<h1>Tutorial 6: Two-Step Refinement Process<a class="headerlink" href="#Tutorial-6:-Two-Step-Refinement-Process" title="Permalink to this heading">¶</a></h1>
<p>Often times,we will want to use <code class="docutils literal notranslate"><span class="pre">Chromo</span></code> to evaluate full-chromosome simulations. This can be computationally expensive and prone to frustration. To mitigate this, we can use a two-step refinement process. In the first step, we can use a coarse-grained model to evaluate the full-chromosome simulation and get an approximate location of large chromatin segments. We then refine the resolution of the coarse-grained model and re-equilibrate the system. This tutorial will demonstrate how to use a
two-step refinement process to evaluate full-chromosome simulations.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="Import-Modules">
<h2>Import Modules<a class="headerlink" href="#Import-Modules" title="Permalink to this heading">¶</a></h2>
<p>The coarse-graining and refining processes are handled by the <code class="docutils literal notranslate"><span class="pre">chromo.util.rediscretize</span></code> module.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Built-in modules</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">inspect</span> <span class="kn">import</span> <span class="n">getmembers</span><span class="p">,</span> <span class="n">isfunction</span>

<span class="c1"># Third-party modules</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1"># Custom modules</span>
<span class="kn">from</span> <span class="nn">chromo.binders</span> <span class="kn">import</span> <span class="n">get_by_name</span><span class="p">,</span> <span class="n">make_binder_collection</span>
<span class="kn">from</span> <span class="nn">chromo.polymers</span> <span class="kn">import</span> <span class="n">Chromatin</span>
<span class="kn">from</span> <span class="nn">chromo.fields</span> <span class="kn">import</span> <span class="n">UniformDensityField</span>
<span class="kn">import</span> <span class="nn">chromo.mc</span> <span class="k">as</span> <span class="nn">mc</span>
<span class="kn">import</span> <span class="nn">chromo.mc.mc_controller</span> <span class="k">as</span> <span class="nn">ctrl</span>
<span class="kn">from</span> <span class="nn">chromo.util.reproducibility</span> <span class="kn">import</span> <span class="n">get_unique_subfolder_name</span>
<span class="kn">import</span> <span class="nn">chromo.util.mu_schedules</span> <span class="k">as</span> <span class="nn">ms</span>
<span class="kn">import</span> <span class="nn">chromo.util.rediscretize</span> <span class="k">as</span> <span class="nn">rd</span>    <span class="c1"># Handles the coarse-graining and refining processes</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Specify-Binders">
<h2>Specify Binders<a class="headerlink" href="#Specify-Binders" title="Permalink to this heading">¶</a></h2>
<p>For this demonstration, we will include two epigenetic marks (H3K9me3 and H3K27me3) and their binders (HP1 and PRC1).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Instantiate HP1 and PRC1 reader proteins</span>
<span class="n">hp1</span> <span class="o">=</span> <span class="n">get_by_name</span><span class="p">(</span><span class="s1">&#39;HP1&#39;</span><span class="p">)</span>
<span class="n">prc1</span> <span class="o">=</span> <span class="n">get_by_name</span><span class="p">(</span><span class="s1">&#39;PRC1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set the HP1-PRC1 cross interaction to 1.0 kT</span>
<span class="n">hp1</span><span class="o">.</span><span class="n">cross_talk_interaction_energy</span><span class="p">[</span><span class="s2">&quot;PRC1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>

<span class="c1"># Set the HP1 and PRC1 chemical potentials to -1.2 kT</span>
<span class="n">hp1</span><span class="o">.</span><span class="n">chemical_potential</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.2</span>
<span class="n">prc1</span><span class="o">.</span><span class="n">chemical_potential</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.2</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a binder collection with the HP1 and PRC1 reader proteins</span>
<span class="n">binder_collection</span> <span class="o">=</span> <span class="n">make_binder_collection</span><span class="p">([</span><span class="n">hp1</span><span class="p">,</span> <span class="n">prc1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Specify-the-Confinement">
<h2>Specify the Confinement<a class="headerlink" href="#Specify-the-Confinement" title="Permalink to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">confine_type</span> <span class="o">=</span> <span class="s2">&quot;Spherical&quot;</span>
<span class="n">confine_radius</span> <span class="o">=</span> <span class="mf">200.0</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Instantiate-Epigenetic-Mark-Pattern">
<h2>Instantiate Epigenetic Mark Pattern<a class="headerlink" href="#Instantiate-Epigenetic-Mark-Pattern" title="Permalink to this heading">¶</a></h2>
<p>When we define the epigenetic mark pattern, we do so at the <strong>full resolution</strong> that we want to obtain following refinement. The epigenetic mark pattern will be rediscretized to generate the coarse-grained mark pattern.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">num_beads</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">mark_pattern</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_beads</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
<span class="n">mark_names</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;H3K9me3&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3&quot;</span><span class="p">])</span>
<span class="n">mark_pattern</span><span class="p">[:</span><span class="mi">500</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">mark_pattern</span><span class="p">[</span><span class="mi">500</span><span class="p">:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Shape of mark pattern:&quot;</span><span class="p">,</span> <span class="n">mark_pattern</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Shape of mark pattern: (1000, 2)
</pre></div></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Specify-Initial-Reader-Protein-Binding-States">
<h2>Specify Initial Reader Protein Binding States<a class="headerlink" href="#Specify-Initial-Reader-Protein-Binding-States" title="Permalink to this heading">¶</a></h2>
<p>Again, we specify the initial reader protein binding states at the <strong>full resolution</strong> that we want to obtain following refinement. The reader protein binding states will be rediscretized to generate the coarse-grained binding states.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">binder_names</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">hp1</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">prc1</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>
<span class="n">states</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_beads</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Instantiate-the-Polymer">
<h2>Instantiate the Polymer<a class="headerlink" href="#Instantiate-the-Polymer" title="Permalink to this heading">¶</a></h2>
<p>We will instantiate the polymer as a <code class="docutils literal notranslate"><span class="pre">Chromatin</span></code> object at full resolution.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the name, number of beads, and bead spacing of the chromatin fiber</span>
<span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;Chr&quot;</span>
<span class="n">bead_spacing</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_beads</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mf">16.5</span>

<span class="c1"># Instantiate the chromatin fiber</span>
<span class="n">poly</span> <span class="o">=</span> <span class="n">Chromatin</span><span class="o">.</span><span class="n">confined_gaussian_walk</span><span class="p">(</span>
    <span class="n">name</span><span class="p">,</span>
    <span class="n">num_beads</span><span class="p">,</span>
    <span class="n">bead_spacing</span><span class="p">,</span>
    <span class="n">confine_type</span><span class="o">=</span><span class="n">confine_type</span><span class="p">,</span>
    <span class="n">confine_length</span><span class="o">=</span><span class="n">confine_radius</span><span class="p">,</span>
    <span class="n">states</span><span class="o">=</span><span class="n">states</span><span class="p">,</span>
    <span class="n">binder_names</span><span class="o">=</span><span class="n">binder_names</span><span class="p">,</span>
    <span class="n">chemical_mods</span><span class="o">=</span><span class="n">mark_pattern</span><span class="p">,</span>
    <span class="n">chemical_mod_names</span><span class="o">=</span><span class="n">mark_names</span><span class="p">,</span>
    <span class="n">max_binders</span><span class="o">=</span><span class="mi">2</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Specify-the-Uniform-Density-Field">
<h2>Specify the Uniform Density Field<a class="headerlink" href="#Specify-the-Uniform-Density-Field" title="Permalink to this heading">¶</a></h2>
<p>The uniform density field is originally specified based on the <strong>full resolution</strong> chromatin fiber.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> <span class="c1"># Specify the dimensions of the field</span>
<span class="n">n_accessible</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">((</span><span class="mi">63</span> <span class="o">*</span> <span class="n">confine_radius</span><span class="p">)</span> <span class="o">/</span> <span class="mi">900</span><span class="p">))</span>
<span class="n">n_buffer</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">n_bins_x</span> <span class="o">=</span> <span class="n">n_accessible</span> <span class="o">+</span> <span class="n">n_buffer</span>
<span class="n">x_width</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">confine_radius</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">n_buffer</span><span class="o">/</span><span class="n">n_accessible</span><span class="p">)</span>
<span class="n">n_bins_y</span> <span class="o">=</span> <span class="n">n_bins_x</span>
<span class="n">y_width</span> <span class="o">=</span> <span class="n">x_width</span>
<span class="n">n_bins_z</span> <span class="o">=</span> <span class="n">n_bins_x</span>
<span class="n">z_width</span> <span class="o">=</span> <span class="n">x_width</span>

<span class="c1"># Initialize the uniform density field</span>
<span class="n">udf</span> <span class="o">=</span> <span class="n">UniformDensityField</span><span class="p">(</span>
    <span class="p">[</span><span class="n">poly</span><span class="p">],</span>
    <span class="n">binder_collection</span><span class="p">,</span>
    <span class="n">x_width</span><span class="p">,</span>
    <span class="n">n_bins_x</span><span class="p">,</span>
    <span class="n">y_width</span><span class="p">,</span>
    <span class="n">n_bins_y</span><span class="p">,</span>
    <span class="n">z_width</span><span class="p">,</span>
    <span class="n">n_bins_z</span><span class="p">,</span>
    <span class="n">confine_type</span><span class="o">=</span><span class="n">confine_type</span><span class="p">,</span>
    <span class="n">confine_length</span><span class="o">=</span><span class="n">confine_radius</span><span class="p">,</span>
    <span class="n">chi</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">assume_fully_accessible</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">fast_field</span><span class="o">=</span><span class="mi">0</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Specify-the-Coarse-Grained-Model">
<h2>Specify the Coarse-Grained Model<a class="headerlink" href="#Specify-the-Coarse-Grained-Model" title="Permalink to this heading">¶</a></h2>
<p>We then coarse-grain the polymer and the field by reducing the number of beads by a fixed factor. In this demonstration, we will reduce the number of beads by a factor of 10.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the factor by which to coarse-grain the polymer and field</span>
<span class="n">cg_factor</span> <span class="o">=</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
<p>There are two ways to generate the epigenetic mark pattern in the coarse-grained chromatin fiber:</p>
<p><strong>(1)</strong> One method involves selecting random epigenetic mark states from fixed-length, non-overlapping segments along the chromatin fiber; stated differently, for each coarse-grained bead, we specify the bead’s mark identity by randomly selecting an epigenetic state from the full-resolution beads represented by it. To initialize epigenetic mark patterns in this way, specify <code class="docutils literal notranslate"><span class="pre">random_states=True</span></code> when coarse-graining the chromatin fiber.</p>
<p><strong>(2)</strong> The other way involves selecting the most common epigenetic mark state from fixed-length, non-overlapping segments along the chromatin fiber; stated differently, for each coarse-grained bead, we specify the bead’s mark identity by selecting the most common epigenetic state from the full-resolution beads represented by it. To initialize epigenetic mark patterns in this way, specify <code class="docutils literal notranslate"><span class="pre">random_states=False</span></code> when coarse-graining the chromatin fiber, or do not specify <code class="docutils literal notranslate"><span class="pre">random_states</span></code> at
all.</p>
<p>When evaluated chromatin structure using an ensemble of coarse-grained representations, the first method does a better job capturing the average behavior of each chromatin segment represented by the coarse-grained beads. To avoid running replicate simulations, we will proceed with the second method.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">poly_cg</span> <span class="o">=</span> <span class="n">rd</span><span class="o">.</span><span class="n">get_cg_chromatin</span><span class="p">(</span>
    <span class="n">polymer</span><span class="o">=</span><span class="n">poly</span><span class="p">,</span>
    <span class="n">cg_factor</span><span class="o">=</span><span class="n">cg_factor</span><span class="p">,</span>
    <span class="n">name_cg</span><span class="o">=</span><span class="s2">&quot;Chr_CG&quot;</span>
<span class="p">)</span>
<span class="n">udf_cg</span> <span class="o">=</span> <span class="n">rd</span><span class="o">.</span><span class="n">get_cg_udf</span><span class="p">(</span>
    <span class="n">udf_refined_dict</span><span class="o">=</span><span class="n">udf</span><span class="o">.</span><span class="n">dict_</span><span class="p">,</span>
    <span class="n">binders_refined</span><span class="o">=</span><span class="n">binder_collection</span><span class="p">,</span>
    <span class="n">cg_factor</span><span class="o">=</span><span class="n">cg_factor</span><span class="p">,</span>
    <span class="n">polymers_cg</span><span class="o">=</span><span class="p">[</span><span class="n">poly_cg</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">binders_cg</span> <span class="o">=</span> <span class="n">rd</span><span class="o">.</span><span class="n">get_cg_binders</span><span class="p">(</span>
    <span class="n">binders_refined</span><span class="o">=</span><span class="n">binder_collection</span><span class="p">,</span>
    <span class="n">cg_factor</span><span class="o">=</span><span class="n">cg_factor</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Specify-the-Simulation-Parameters">
<h2>Specify the Simulation Parameters<a class="headerlink" href="#Specify-the-Simulation-Parameters" title="Permalink to this heading">¶</a></h2>
<p>For the purposes of this demonstration, I will significantly reduce the number of mc steps per snapshot. This will allow us to run the simulation faster. <strong>The simulations will not be equilibrated!</strong> For more rigorous simulations, increase the number of mc steps per snapshot to ensure that the system reaches equilibrium.</p>
<p>We specify the simulation parameters based on the <strong>coarse-grained</strong> chromatin fiber at this step.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">amp_bead_bounds</span><span class="p">,</span> <span class="n">amp_move_bounds</span> <span class="o">=</span> <span class="n">mc</span><span class="o">.</span><span class="n">get_amplitude_bounds</span><span class="p">(</span>
    <span class="n">polymers</span> <span class="o">=</span> <span class="p">[</span><span class="n">poly_cg</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the number of snapshots and the number of MC steps to attempt per snapshot</span>
<span class="n">num_snapshots</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">mc_steps_per_snapshot</span> <span class="o">=</span> <span class="mi">10</span>  <span class="c1"># Reduce the number of MC steps per snapshot to speed up the simulation</span>
<span class="c1"># TODO: If you want to run a more rigorous simulation, increase the number of MC steps per snapshot</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a list of simulated annealing schedules, which are defined in another file</span>
<span class="n">schedules</span> <span class="o">=</span> <span class="p">[</span><span class="n">func</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">func</span> <span class="ow">in</span> <span class="n">getmembers</span><span class="p">(</span><span class="n">ms</span><span class="p">,</span> <span class="n">isfunction</span><span class="p">)]</span>

<span class="c1"># Select a schedule for slowly adding HP1 into the system</span>
<span class="n">select_schedule</span> <span class="o">=</span> <span class="s2">&quot;linear_step_for_negative_cp&quot;</span>
<span class="n">mu_schedules</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">ms</span><span class="o">.</span><span class="n">Schedule</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">ms</span><span class="p">,</span> <span class="n">func_name</span><span class="p">))</span> <span class="k">for</span> <span class="n">func_name</span> <span class="ow">in</span> <span class="n">schedules</span>
<span class="p">]</span>
<span class="n">mu_schedules</span> <span class="o">=</span> <span class="p">[</span><span class="n">sch</span> <span class="k">for</span> <span class="n">sch</span> <span class="ow">in</span> <span class="n">mu_schedules</span> <span class="k">if</span> <span class="n">sch</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">select_schedule</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Run-the-Coarse-Grained-Simulation">
<h2>Run the Coarse-Grained Simulation<a class="headerlink" href="#Run-the-Coarse-Grained-Simulation" title="Permalink to this heading">¶</a></h2>
<p>The output of the simulation will be a coarse-grained representation of the chromatin fiber.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the output directory</span>
<span class="n">out_dir</span> <span class="o">=</span> <span class="s2">&quot;output_demo&quot;</span>

<span class="c1"># Run the simulation</span>
<span class="n">polymers</span> <span class="o">=</span> <span class="n">mc</span><span class="o">.</span><span class="n">polymer_in_field</span><span class="p">(</span>
    <span class="p">[</span><span class="n">poly_cg</span><span class="p">],</span>
    <span class="n">binders_cg</span><span class="p">,</span>
    <span class="n">udf_cg</span><span class="p">,</span>
    <span class="n">mc_steps_per_snapshot</span><span class="p">,</span>
    <span class="n">num_snapshots</span><span class="p">,</span>
    <span class="n">amp_bead_bounds</span><span class="p">,</span>
    <span class="n">amp_move_bounds</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">,</span>
    <span class="n">mu_schedule</span><span class="o">=</span><span class="n">mu_schedules</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Plot-the-Coarse-Grained-Chromatin-Fiber">
<h2>Plot the Coarse-Grained Chromatin Fiber<a class="headerlink" href="#Plot-the-Coarse-Grained-Chromatin-Fiber" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the final configuration of the polymer</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">polymers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">r</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">polymers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">r</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">z</span> <span class="o">=</span> <span class="n">polymers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">r</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">projection</span><span class="o">=</span><span class="s1">&#39;3d&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot3D</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">y</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">z</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;x&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;y&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="s1">&#39;z&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_6_two_step_refinement_28_0.png" src="_images/tutorial_6_two_step_refinement_28_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of Beads in the Coarse-Grained Chromatin Fiber: </span><span class="si">{</span><span class="n">polymers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_beads</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of Beads in the Coarse-Grained Chromatin Fiber: 100
</pre></div></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Refine-the-Coarse-Grained-Model">
<h2>Refine the Coarse-Grained Model<a class="headerlink" href="#Refine-the-Coarse-Grained-Model" title="Permalink to this heading">¶</a></h2>
<p>With the coarse-grained simulation equilibrated, we have an approximation of where genomic segments lie in space, relative to one-another. We can now insert beads back into the model and re-equilibrate the configuration to get a higher-resolution representation of the chromatin fiber. While we know the epigenetic mark pattern for the refined model (since it was specified prior to coarse-graining), we do not know the reader protein binding states. We allow a certain number of MC steps to
equilibrate the binding states prior to proposing geometric moves to the refined chromatin fiber.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Important-Consideration">
<h2>Important Consideration<a class="headerlink" href="#Important-Consideration" title="Permalink to this heading">¶</a></h2>
<p>As of July 3, 2024, only uniform linker lengths are supported by the refinement step. This can be modified during future development.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify how many binding moves to attempt prior to proposing geometric moves</span>
<span class="c1"># This should be increased (e.g., to 1000000) for a more rigorous simulation</span>
<span class="n">n_bind_eq</span> <span class="o">=</span> <span class="mi">100</span>

<span class="c1"># Specify the spacing between beads in the refined chromatin fiber</span>
<span class="n">bead_spacing</span> <span class="o">=</span> <span class="mf">16.5</span>  <span class="c1"># Currently, only uniform linker lengths are supported</span>

<span class="c1"># Refine the coarse-grained model</span>
<span class="n">poly_refine</span><span class="p">,</span> <span class="n">udf_refine</span> <span class="o">=</span> <span class="n">rd</span><span class="o">.</span><span class="n">refine_chromatin</span><span class="p">(</span>
    <span class="n">polymer_cg</span><span class="o">=</span><span class="n">polymers</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>    <span class="c1"># This comes from the output of the coarse-grained model</span>
    <span class="n">num_beads_refined</span><span class="o">=</span><span class="n">num_beads</span><span class="p">,</span>
    <span class="n">bead_spacing</span><span class="o">=</span><span class="n">bead_spacing</span><span class="p">,</span>
    <span class="n">chemical_mods</span><span class="o">=</span><span class="n">mark_pattern</span><span class="p">,</span>
    <span class="n">udf_cg</span><span class="o">=</span><span class="n">udf_cg</span><span class="p">,</span>
    <span class="n">binding_equilibration</span><span class="o">=</span><span class="n">n_bind_eq</span><span class="p">,</span>
    <span class="n">name_refine</span><span class="o">=</span><span class="s2">&quot;Chr_refine&quot;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="n">out_dir</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id1">
<h2>Specify the Simulation Parameters<a class="headerlink" href="#id1" title="Permalink to this heading">¶</a></h2>
<p>Now, re-specify simulation parameters based on the refined polymer. Notice, you do not want to apply simulated annealing to reader protein chemical potentials. If the approximate configuration generated by the coarse step is lost due to a large change in chemical potential, then the two-step refinement process is not useful.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify move and bead amplitudes</span>
<span class="n">amp_bead_bounds</span><span class="p">,</span> <span class="n">amp_move_bounds</span> <span class="o">=</span> <span class="n">mc</span><span class="o">.</span><span class="n">get_amplitude_bounds</span><span class="p">([</span><span class="n">poly_refine</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the number of snapshots and the number of MC steps to attempt per snapshot</span>
<span class="n">num_snapshots</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">mc_steps_per_snapshot</span> <span class="o">=</span> <span class="mi">10</span>  <span class="c1"># Reduce the number of MC steps per snapshot to speed up the simulation</span>
<span class="c1"># TODO: If you want to run a more rigorous simulation, increase the number of MC steps per snapshot</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Run-the-Refined-Simulation">
<h2>Run the Refined Simulation<a class="headerlink" href="#Run-the-Refined-Simulation" title="Permalink to this heading">¶</a></h2>
<p>The output of the simulation will be a refined representation of the chromatin fiber.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run the refined simulation</span>
<span class="n">polymers_refined</span> <span class="o">=</span> <span class="n">mc</span><span class="o">.</span><span class="n">polymer_in_field</span><span class="p">(</span>
    <span class="p">[</span><span class="n">poly_refine</span><span class="p">],</span>
    <span class="n">binder_collection</span><span class="p">,</span>    <span class="c1"># This was specified prior to coarse-graining</span>
    <span class="n">udf_refine</span><span class="p">,</span>
    <span class="n">mc_steps_per_snapshot</span><span class="p">,</span>
    <span class="n">num_snapshots</span><span class="p">,</span>
    <span class="n">amp_bead_bounds</span><span class="p">,</span>
    <span class="n">amp_move_bounds</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="Plot-the-Refined-Chromatin-Fiber">
<h2>Plot the Refined Chromatin Fiber<a class="headerlink" href="#Plot-the-Refined-Chromatin-Fiber" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the final configuration of the polymer</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">polymers_refined</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">r</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">polymers_refined</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">r</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">z</span> <span class="o">=</span> <span class="n">polymers_refined</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">r</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">projection</span><span class="o">=</span><span class="s1">&#39;3d&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot3D</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">y</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">z</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;x&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;y&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="s1">&#39;z&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_6_two_step_refinement_39_0.png" src="_images/tutorial_6_two_step_refinement_39_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of Beads in the Refined Chromatin Fiber: </span><span class="si">{</span><span class="n">polymers_refined</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_beads</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of Beads in the Refined Chromatin Fiber: 1000
</pre></div></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="tutorial_7_kinked_chromatin.html" class="btn btn-neutral float-right" title="Tutorial 7: Kinked Chromatin" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="tutorial_5_two_mark_chromatin_competitive.html" class="btn btn-neutral" title="Tutorial 5: Two-Mark Chromatin with Competitive Binding" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2024, Joseph Wakim, Bruno Beltran, Andrew Spakowitz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> and ❤️  using a custom <a href="https://github.com/LinxiFan/Sphinx-theme">theme</a> based on <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/_sphinx_javascript_frameworks_compat.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="_static/sphinx_highlight.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script type="text/javascript" src="None"></script>
      <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>